[kubernetes][autoscaler] Kubernetes operator basic fixes

DmitriGekhtman opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This PR makes several low-effort fixes to the Kubernetes Operator

- Restores setup commands for parity with current cluster launcher
- Adds the ability to run a cluster-scoped operator. 
- Backs the operator by a one-pod deployment. (Not a theoretical issue; have gotten some user complaints.)
- If the operator is created before the CRD, have it wait for the CRD to be created
- The head service selector is added to the head pod before launch, instead of being added to head pod type
- Automatically prepend the name of the cluster to each of the pods
- Garbage collect the head service upon cluster deletion
- When running the operator, skip the bootstrap steps that try to give the head pod autoscaler permissions (these produce extraneous messages in the logs)
- Changes foo, bar, baz, quux names of example custom resources to something more intelligible.

Also, a small modification to a method of the Monitor: instead of using `args.redis_password` and `args.redis_address`, make these attributes of the Monitor object and use those attributes.
(The operator doesn't instantiate Monitor from command line, so `args` is undefined.) 

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

Closes #14726
Closes #14093
Closes #14575
Addresses #14700 for Operator

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(
  
Tests of the following added:
- Cluster-scoped operator works
- Scaling functionality survives when operator is deleted and brought back by Deployment controller
- Things work fine if CRD is created after operator starts
- Head service is garbage-collected

DmitriGekhtman commented:
need to identify Ray clusters by name and namespace in the code
`ray_clusters[(name, namespace)]`
rather than just
`ray_clusters[name]`

yiranwang52 commented on python/ray/autoscaler/_private/_kubernetes/node_provider.py:
```diff
@@ -10,14 +11,22 @@
+def head_service_selector(cluster_name: str) -> Dict[str, str]:
+    """Selector for Operator-configured head service.
+    """
+    return {"cluster.ray.io/component": f"{cluster_name}-ray-head"}
```
should `cluster.ray.io/component` be a constant?

DmitriGekhtman replied:
Yes.

yiranwang52 commented on python/ray/ray_operator/operator.py:
```diff
@@ -141,23 +158,20 @@ def cluster_action(event_type, cluster_cr, cluster_name) -> None:
+                logger.warning("Waiting for creation of the RayCluster CRD")
+                time.sleep(5)
+            else:
+                raise
```
what happens after this?

DmitriGekhtman replied:
That's an unhandled exception...
It would mean we failed to enter the event loop for some unforeseen reason.
The container crashes within a short period of starting, then restarts. 

If we got to this line of code on the first try, then probably we'll quickly get to it on the second try too. Eventually we'd get CrashLoopBackOff and could check what's wrong with kubectl logs.

DmitriGekhtman replied:
Umm, same behavior if the exception raised is not an ApiException.

yiranwang52 replied:
okay crash looping makes sense.

DmitriGekhtman commented:
Going to test again. 
Will add check that launching with same names in different namespaces isn't a problem.

DmitriGekhtman commented:
Updated everything where Ray clusters were identified by name -- they're now identified by name and namespace.
Added logic to the tests to check that creating a clusters with the same name in different namespaces is fine. Tests pass.

This PR is ready for review again.

yiranwang52 commented:
I am surprised `_generation` needs to be handled manually

edoakes commented:
Trusting tests-ok label from @DmitriGekhtman
